helix-core/src/main/java/org/apache/helix/model/TrieClusterTopology.java

```diff
@@ -0,0 +1,340 @@
+   * @param node the node from where to collect all the nodes' paths.
+   * @return All the paths under the node.
+   */
+  private Set<String> getPathUnderNode(TrieNode node) {
+    Set<String> resultMap = new HashSet<>();
+    Deque<TrieNode> nodeStack = new ArrayDeque<>();
+    nodeStack.push(node);
+    while (!nodeStack.isEmpty()) {
+      node = nodeStack.pop();
+      if (node.getChildren().isEmpty()) {
+        resultMap.add(node.getPath());
+      } else {
+        for (TrieNode child : node.getChildren().values()) {
+          nodeStack.push(child);
+        }
+      }
+    }
+    return resultMap;
+  }
+
+  private TrieNode getStartNode(LinkedHashMap<String, String> domain) {
+    TrieNode curNode = _rootNode;
+    TrieNode nextNode;
+    for (Map.Entry<String, String> entry : domain.entrySet()) {
+      nextNode = curNode.getChildren().get(entry.getKey() + CONNECTOR + entry.getValue());
+      if (nextNode == null) {
+        throw new IllegalArgumentException(String
+            .format("The input domain %s does not have the value %s", entry.getKey(),
+                entry.getValue()));
+      }
+      curNode = nextNode;
+    }
+    return curNode;
+  }
+
+  private List<TrieNode> getStartNodes(String domain) {
+    List<TrieNode> results = new ArrayList<>();
+    TrieNode curNode = _rootNode;
+    Deque<TrieNode> nodeStack = new ArrayDeque<>();
+    nodeStack.push(curNode);
+    while (!nodeStack.isEmpty()) {
+      curNode = nodeStack.pop();
+      if (curNode.getDomainType().equals(domain)) {
+        results.add(curNode);
+      } else {
+        for (TrieNode child : curNode.getChildren().values()) {
+          nodeStack.push(child);
+        }
+      }
+    }
+    return results;
+  }
+
+  private void removeInvalidInstanceConfig(Map<String, InstanceConfig> instanceConfigMap) {
+    Set<String> toRemoveConfig = new HashSet<>();
+    for (String instanceName : instanceConfigMap.keySet()) {
+      Map<String, String> domainAsMap = instanceConfigMap.get(instanceName).getDomainAsMap();
+      if (domainAsMap.isEmpty()) {
+        logger.info(String.format("Domain for instance %s is not set", instanceName));
+        toRemoveConfig.add(instanceName);
+      } else {
+        for (String key : _topologyKeys) {
+          String value = domainAsMap.get(key);
+          if (value == null || value.length() == 0) {
+            logger.info(String.format("Domain %s for instance %s is not set", domainAsMap.get(key),
+                instanceName));
+            toRemoveConfig.add(instanceName);
+            break;
+          }
+        }
+      }
+    }
+    toRemoveConfig.forEach(entry -> instanceConfigMap.remove(entry));
+  }
+
+  /**
+   * Constructs a trie based on the provided instance config map. It loops through all instance
+   * configs and constructs the trie in a top down manner.
+   */
+  private void constructTrie(Map<String, InstanceConfig> instanceConfigMap) {
+    removeInvalidInstanceConfig(instanceConfigMap);
```

Comment:
How are the callers supposed to know which instance has been removed because the configuration is invalid?
In addition, the removal is directly modifying the input map. So it could dangerous since it may change the caller's map implicitly.

Reply:
Please see the updated PR. I add an invalid config list that customers can retrieve.